src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SymbolToProjectId.cs

```diff
@@ -0,0 +1,135 @@
+                = new ConditionalWeakTable<Compilation, ProjectId>();
+
+            private readonly SolutionState _solutionState;
+
+            /// <summary>
+            /// Cache we use to map between assembly and module symbols and the project they came from.  That way if we
+            /// are asked about many symbols from the same assembly/module we can answer the question quickly after
+            /// computing for the first one.
+            /// </summary>
+            private ConditionalWeakTable<ISymbol, ProjectId?>? _assemblyOrModuleSymbolToProjectId;
+            private static readonly Func<ConditionalWeakTable<ISymbol, ProjectId?>> s_createTable = () => new ConditionalWeakTable<ISymbol, ProjectId?>();
+
+            public SymbolToProjectId(SolutionState solutionState)
+            {
+                _solutionState = solutionState;
+                _assemblyOrModuleSymbolToProjectId = null;
+            }
+
+            /// <inheritdoc cref="Solution.GetExactProjectId(ISymbol)"/>.
+            public ProjectId? GetExactProjectId(ISymbol? symbol)
+            {
+                LazyInitialization.EnsureInitialized(ref _assemblyOrModuleSymbolToProjectId, s_createTable);
+
+                // Walk up the symbol so we can get to the containing namespace/assembly that will be used to map
+                // back to a project.
+
+                while (symbol != null)
+                {
+                    var result = GetProjectIdDirectly(symbol, _assemblyOrModuleSymbolToProjectId);
+                    if (result != null)
+                        return result;
+
+                    symbol = symbol.ContainingSymbol;
+                }
+
+                return null;
+            }
+
+            private ProjectId? GetProjectIdDirectly(
+                ISymbol symbol, ConditionalWeakTable<ISymbol, ProjectId?> assemblyOrModuleSymbolToProjectId)
+            {
+                if (symbol.IsKind(SymbolKind.Namespace, out INamespaceSymbol? ns))
+                {
+                    if (ns.ContainingCompilation != null)
+                    {
+                        // A namespace that spans a compilation.  These don't belong to an assembly/module. However, we
+                        // can map their compilation directly 
+                        s_compilationToProjectId.TryGetValue(ns.ContainingCompilation, out var projectId);
+                        return projectId;
+                    }
+                }
+                else if (symbol.IsKind(SymbolKind.Assembly) ||
+                         symbol.IsKind(SymbolKind.NetModule))
+                {
+                    if (!assemblyOrModuleSymbolToProjectId.TryGetValue(symbol, out var projectId))
+                    {
+                        foreach (var (id, state) in _solutionState.ProjectStates)
+                        {
+                            var tracker = _solutionState.GetCompilationTracker(id);
```

Comment:
This probably should just directly enumerate _projectIdToTrackerMap, since this is otherwise going to realize CompilationTrackers for projects that didn't have one (and then definitely couldn't have been the source of the symbols, I think?)

Reply:
fair point :)

Reply:
done!